docker/docker_image_src.go

```diff
@@ -30,15 +32,73 @@ type dockerImageSource struct {
 
 // newImageSource creates a new ImageSource for the specified image reference.
 // The caller must call .Close() on the returned ImageSource.
-func newImageSource(sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
-	c, err := newDockerClientFromRef(sys, ref, false, "pull")
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	registry, err := sysregistriesv2.FindRegistry(sys, ref.ref.Name())
+	if err != nil {
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+
+	var imageSource *dockerImageSource
+	if registry == nil {
+		// If no registry was found during sysregistriesv2 lookup, we fallback
+		// to the standard behavior in pulling the fully qualified image from
+		// the remote.
+		c, err := newDockerClientFromRef(sys, ref, false, "pull")
+		if err != nil {
+			return nil, err
+		}
+
+		imageSource = &dockerImageSource{ref: ref, c: c}
+		if err := imageSource.ensureManifestIsLoaded(ctx); err != nil {
+			return nil, err
+		}
+
+		return imageSource, nil
+	}
+
+	// Found the registry within the sysregistriesv2 configuration. Now we test
+	// all endpoints for the manifest availability. If an working image source
+	// was found, it will be used for all future pull actions.
+	for _, endpoint := range append(registry.Mirrors, registry.Endpoint) {
+		newRef, err := rewriteReference(ref, registry.Prefix, endpoint.URL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error rewriting reference")
+		}
+
+		client, err := newDockerClientFromRef(sys, *newRef, false, "pull")
+		client.endpointSkipVerify = endpoint.Insecure
+		if err != nil {
+			return nil, err
+		}
+
+		testImageSource := &dockerImageSource{
+			ref: *newRef,
+			c:   client,
+		}
+
+		if testImageSource.ensureManifestIsLoaded(ctx) == nil {
+			imageSource = testImageSource
+			break
+		}
+	}
+
+	if imageSource == nil {
+		return nil, fmt.Errorf("unable to determine registry endpoint")
+	}
+
+	return imageSource, nil
+}
+
+// rewriteReference will substitute the provided registry `prefix` with the
+// base `url` from the `ref` and creates a new reference from it. The function
+// errors if the newly created reference is not parsable.
+func rewriteReference(ref dockerReference, prefix, url string) (*dockerReference, error) {
+	newRegURL := strings.Replace(ref.ref.String(), prefix, url, 1)
```

Comment:
- This must _only_ match at the start of the string.
- `dockerReference` is usually passed by value, to avoid unexpected sharing (and because it only contains a single interface pointer, an extra level of indirection is not making passing the values around noticeably more efficient anyway.)
- (None of these values are URLs. Maybe something vaguely like (from, to) instead of (prefix, url)?)

Reply:
> * This must _only_ match at the start of the string.

Sure, I added an additional check for that.

> * `dockerReference` is usually passed by value, to avoid unexpected sharing (and because it only contains a single interface pointer, an extra level of indirection is not making passing the values around noticeably more efficient anyway.)

Changed it according your comment.

> * (None of these values are URLs. Maybe something vaguely like (from, to) instead of (prefix, url)?)

Makes sense, I changed them as suggested by you.